dotCMS/src/main/java/com/dotmarketing/portlets/containers/action/EditContainerAction.java

```diff
@@ -559,43 +563,78 @@ public void _saveWebAsset(ActionRequest req, ActionResponse res,
 		identifier.setHostId(host.getIdentifier());
 		APILocator.getIdentifierAPI().save(identifier);
 
+
+		//Save/Update/Delete the ContainerStructures associated
+		saveContainerStructures(req, currentContainer, container);
+
+
+		SessionMessages.add(httpReq, "message", "message.containers.save");
+		ActivityLogger.logInfo(this.getClass(), "Save WebAsset action", "User " + user.getPrimaryKey() + " saved " + container.getTitle(), HostUtil.hostNameUtil(req, _getUser(req)));
+		// saves to working folder under velocity
+		ContainerServices.invalidate(container, true);
+
+		// copies the information back into the form bean
+		BeanUtils.copyProperties(form, req
+				.getAttribute(WebKeys.CONTAINER_FORM_EDIT));
+
+		APILocator.getVersionableAPI().setWorking(container);
+
+
+
+
+		HibernateUtil.flush();
+	}
+
+	/**
+	 * Containers are associated with multiple Structures. This method takes care of the ContainerStructure save/update/delete logic
+	 * @param req Request
+	 * @param currentContainer Container as currently exists
+	 * @param container Container with the new changes to be persisted
+	 */
+	private void saveContainerStructures (ActionRequest req, Container currentContainer, Container container)
+			throws DotDataException, DotSecurityException {
```

Comment:
The logic on this code can be optimized keeping a map for oldContainerStructures, then filter to exclude newContainerStructures (https://www.mkyong.com/java8/java-8-filter-a-map-examples/). With the resulting map you can delete in batch. I think it's a faster and less verbose implementation